lib/grandpa/finalization.go

```diff
@@ -0,0 +1,493 @@
+		if err != nil {
+			fh.observableErrs <- fmt.Errorf("initiating round: %w", err)
+			return
+		}
+
+		err = fh.waitServices()
+		if err != nil {
+			fh.observableErrs <- err
+			return
+		}
+	}
+}
+
+func (fh *finalizationHandler) stop() (err error) {
+	fh.servicesLock.Lock()
+	defer fh.servicesLock.Unlock()
+
+	stopWg := new(sync.WaitGroup)
+	stopWg.Add(2)
+
+	errs := [2]error{}
+	go func() {
+		defer stopWg.Done()
+		errs[0] = fh.finalizationEngine.Stop()
+	}()
+
+	go func() {
+		defer stopWg.Done()
+		errs[1] = fh.votingRound.Stop()
+	}()
+
+	stopWg.Wait()
+
+	if errs[0] != nil {
+		return errs[0]
+	}
+
+	if errs[1] != nil {
+		return errs[1]
+	}
+
+	return nil
+}
+
+func (fh *finalizationHandler) Stop() (err error) {
+	close(fh.stopCh)
+
+	err = fh.stop()
+	select {
+	case <-fh.handlerDone:
+	case <-time.After(fh.timeoutStop):
+		return errTimeoutWhileStoping
+	}
+
+	close(fh.observableErrs)
+	return err
+}
+
+// waitServices will start the services and wait until they complete or
+func (fh *finalizationHandler) waitServices() error {
+	fh.servicesLock.Lock()
+	fh.finalizationEngine, fh.votingRound = fh.newServices()
+	fh.servicesLock.Unlock()
+
+	finalizationEngineErr := make(chan error)
+	go func() {
+		defer close(finalizationEngineErr)
+		err := fh.finalizationEngine.Start()
+		if err != nil {
+			finalizationEngineErr <- err
+		}
+	}()
+
+	votingRoundErr := make(chan error)
+	go func() {
+		defer close(votingRoundErr)
+		err := fh.votingRound.Start()
+		if err != nil {
+			votingRoundErr <- err
+		}
+	}()
+
+	for {
+		select {
+		case <-fh.stopCh:
+			return nil
+
+		case err, ok := <-votingRoundErr:
+			if !ok {
+				votingRoundErr = nil
+			}
+
+			stopErr := fh.stop()
+			if stopErr != nil {
+				logger.Infof("stopping finalisation handler: %s", stopErr)
+			}
+			return err
+
+		case err, ok := <-finalizationEngineErr:
+			if !ok {
+				finalizationEngineErr = nil
+			}
+
+			stopErr := fh.stop()
+			if stopErr != nil {
+				logger.Infof("stopping finalisation handler: %s", stopErr)
+			}
+
+			return err
+
+		default:
+			finish := votingRoundErr == nil && finalizationEngineErr == nil
+			if finish {
+				return nil
+			}
+		}
+	}
+}
+
+var errTimeoutWhileStoping = errors.New("timeout while stopping")
+
+type handleVotingRound struct {
+	grandpaService       *Service
+	timeoutStop          time.Duration
+	finalizationEngineCh <-chan engineAction
+	stopCh               chan struct{}
+	engineDone           chan struct{}
+}
+
+func newHandleVotingRound(service *Service, finalizationEngineCh <-chan engineAction) *handleVotingRound {
+	return &handleVotingRound{
+		timeoutStop:          5 * time.Second,
+		grandpaService:       service,
+		stopCh:               make(chan struct{}),
+		engineDone:           make(chan struct{}),
+		finalizationEngineCh: finalizationEngineCh,
+	}
+}
+
+func (h *handleVotingRound) Stop() (err error) {
+	if h.stopCh == nil {
+		return nil
+	}
+
+	close(h.stopCh)
+	select {
+	case <-h.engineDone:
```

Comment:
From the top of my head and what I've read yesterday, when you close `stopCh`, it signals the goroutine to stop. However, the goroutine is not instantly closed, so we wait for it to terminate with `engineDone`. There probably is a `defer close(engineDone)` in the goroutine function.

Reply:
Yeah we should probably document this in the eng handbook, this is rather critical (to avoid 'fire and forget' goroutines which are untestable, unpredictable, and may lead to a build up of goroutines). And in my experience so far, not many devs know or think about it (also shame on contexts carrying a mechanism system but not a waiting for canceled mechanism) :confused: